dma: Start/stop behavioral contract

teburd opened the pull request:
Further specifies the behavioral contract for start/stop to be callable regardless of the channel status.

Commit:
dma: Document expected behavior of start/stop

The DMA API expects drivers in effect to maintain their own internal
state of channels. A channel that is already started may have dma_start
called on it again without effect. A channel that is already stopped may
have dma_stop called on it without effect.

In essence, start and stop are expected to act like events a DMA channels
finite state machine reacts to and if the state is already at the desired
one then nothing is to be done and no error has occurred.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
dma: Test repeatedly calling start/stop

Ensures that the documented behavior of the API is met by implementations
through testing. By calling stop on a stopped channel the expectation is no
error occurs and is checked.

Calling start after a channel has been started is difficult to test for
as there is transfer timing involved. A once shot transfer may have
completed and the channel transition to an inactive state by itself by
the time the second start call is performed. This isn't supported by at
least gpdma today and isn't documented behaviorally so should not be
tested.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
dma: HDA ignore repeated start/stop requests

The DMA API contract specifies that start/stop may be called multiple
times. Prior to adding power management this was perfectly fine as it was.
In adding power management, there are additional side effects that can
cause issues. Instead check the state of the channel prior to start/stop
and do nothing if already in the desired state.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

tmleman commented on soc/xtensa/intel_adsp/common/include/intel_adsp_hda.h:
```diff
@@ -210,7 +210,7 @@ static inline uint32_t intel_adsp_hda_get_buffer_size(uint32_t base,
 static inline void intel_adsp_hda_enable(uint32_t base, uint32_t regblock_size, uint32_t sid)
 {
-	*DGCS(base, regblock_size, sid) |= DGCS_GEN | DGCS_FIFORDY;
+	return *DGCS(base, regblock_size, sid) |= DGCS_GEN | DGCS_FIFORDY;
```
This _return_ should be in function \*is_enabled :)

tmleman commented on drivers/dma/dma_intel_adsp_hda.c:
```diff
@@ -243,6 +247,10 @@ int intel_adsp_hda_dma_stop(const struct device *dev, uint32_t channel)
 {
 	const struct intel_adsp_hda_dma_cfg *const cfg = dev->config;
 
+	if (!intel_adsp_hda_is_enabled(cfg->base, cfg->regblock_size, channel)) {
+		return 0;
+	}
```
I would rather put it after the **\_\_ASSEERT**

tmleman commented:
By the way, I noticed that the function documentation is out of date. None of them accept _**hda**_ parameter.

Commit:
soc: intel_adsp: Correct HDA parameter docstrings

The parameter doc string for hda was incorrect as the parameters
had been updated to take the IP base address, block size, and stream id
instead. Updates all doc string comments to account for the change.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
> By the way, I noticed that the function documentation is out of date. None of them accept _**hda**_ parameter.

Fixed, thanks for pointing that out!

tmleman commented on soc/xtensa/intel_adsp/common/include/intel_adsp_hda.h:
```diff
@@ -118,7 +118,7 @@
  * @brief Initialize an HDA stream for use with the firmware
  *
- * @param hda Stream set to work with
+ * @param base Base address of
```
"the IP register block" string is missing.

tmleman commented on soc/xtensa/intel_adsp/common/include/intel_adsp_hda.h:
```diff
@@ -229,7 +228,7 @@ static inline void intel_adsp_hda_disable(uint32_t base, uint32_t regblock_size,
+ * @param base Base address of the IP register block
  * @param regblock_size Register block size
  * @param sid Stream ID
  */
```
This should have been fixed in the previous commit.

teburd commented:
Draft until I fix gpdma to also work with repeated start/stop as it currently fails.

hakehuang commented on drivers/dma/dma_intel_adsp_hda.c:
```diff
@@ -230,6 +230,10 @@ int intel_adsp_hda_dma_start(const struct device *dev, uint32_t channel)
 
 	__ASSERT(channel < cfg->dma_channels, "Channel does not exist");
 
+	if (intel_adsp_hda_is_enabled(cfg->base, cfg->regblock_size, channel)) {
```
I guess this could be a common implement, can we put it to dam.h z_impl_dma_start and z_impl_dma_stop?  otherwise all other venders driver need update so

teburd replied:
Are you looking to have the driver API implement a function that start/stop can inspect then? That probably makes good sense

hakehuang replied:
right, and this API can have a default one, as we call start/stop in dma.h, and we can have internal lock. like below

`
/**
 * DMA context structure
 * Note: the dma_context shall be the first member
 *       of DMA client driver Data, got by dev->data
 *
 * magic			- magic code to identify the context
 * dma_channels		- dma channels
 * atomic			- driver atomic_t pointer
 *
 */
#define MAX_CHANNELS 64
struct dma_context {
	int32_t magic;
	int dma_channels;
	atomic_t *atomic;
       ATOMIC_DEFINE(channels_status_atomic, MAX_CHANNELS);
};

static inline int z_impl_dma_start(const struct device *dev, uint32_t channel)
{
	const struct dma_driver_api *api =
		(const struct dma_driver_api *)dev->api;
        struct dma_context *dma_ctx = (struct dma_context *)dev->data;

        if(atomic_test_and_set_bit(&dma_ctx->channels_status_atomic, channel)){
	        return api->start(dev, channel);
        } else {
                return 0;
        }
}

static inline int z_impl_dma_stop(const struct device *dev, uint32_t channel)
{
	const struct dma_driver_api *api =
		(const struct dma_driver_api *)dev->api;
       struct dma_context *dma_ctx = (struct dma_context *)dev->data;

       if(!atomic_test_and_clear_bit(&dma_ctx->channels_status_atomic, channel))
       {
               return 0;
       } else {
	        return api->stop(dev, channel);
      }
}

`

teburd replied:
Its an interesting idea, though in some cases today repeated start I think may mean different things.

Perhaps starting this out limiting this behavioral change to be repeated stop as being allowable is better and simpler. I believe that resolves the issue @tmleman ran into. It's also easier to test.

Commit:
dma: Repeated start/stop support for gpdma

Allows for repeatedly calling start/stop on designwares gpdma.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

hakehuang commented:
> Draft until I fix gpdma to also work with repeated start/stop as it currently fails the test. Adding other HAL/DMA folks. @erwango @hakehuang let me know if repeated start/stop as a contract is an issue for st or nxp please.

dma_mcux_lpc platfroms has such issue, whereas the dma_mcux_edma does not. but it is better to add protection both. in case HAL code changed.

teburd commented:
@tmleman Is documenting and expecting repeated stop enough or is repeated calls to start also required?

FRASTM commented:
STM32 dma driver has to be updated in the same way:

-  in dma_stm32_start()
```
	if (LL_DMA_IsEnabledChannel(dma, dma_stm32_id_to_stream(id))) {
		return 0;
	}
```

- in dma_stm32_stop()
```
	if (!LL_DMA_IsEnabledChannel(dma, dma_stm32_id_to_stream(id))) {
		return 0;
	}
```

tmleman commented:
> Is documenting and expecting repeated stop enough or is repeated calls to start also required?

@teburd from what I've seen, only stop is triggered several times for the same channel. But form a power manager perspective it is required to ensure that a given channel is started and stopped only once.

teburd commented:
> > Is documenting and expecting repeated stop enough or is repeated calls to start also required?
> 
> @teburd from what I've seen, only stop is triggered several times for the same channel. But form a power manager perspective it is required to ensure that a given channel is started and stopped only once.

Understood, the changeset here should do just that. Testing repeated start though is problematic as the one shot memory to memory transfers being done here may already be done by the time start is called again. I think a state diagram in the docs might help clarify the possible scenarios and is something I've had in my mind here as I'm doing this. Would that be helpful to others?
